diff --git a/include/crt_config.h b/include/crt_config.h
new file mode 100644
index 000000000..912201dce
--- /dev/null
+++ b/include/crt_config.h
@@ -0,0 +1,34 @@
+#ifndef MODULES_TVM_CRT_CONFIG_H_
+#define MODULES_TVM_CRT_CONFIG_H_
+
+/* Log level of the CRT runtime */
+#define TVM_CRT_LOG_LEVEL TVM_CRT_LOG_LEVEL_DEBUG
+
+/* Support low-level debugging in MISRA-C runtime */
+#define TVM_CRT_DEBUG 0
+
+/* Maximum supported dimension in NDArray */
+#define TVM_CRT_MAX_NDIM 4
+
+/* Maximum supported arguments in generated functions */
+#define TVM_CRT_MAX_ARGS 5
+
+/* Size of the global function registry, in bytes. */
+#define TVM_CRT_GLOBAL_FUNC_REGISTRY_SIZE_BYTES 256
+
+/* Maximum number of registered modules. */
+#define TVM_CRT_MAX_REGISTERED_MODULES 1
+
+/* Maximum packet size, in bytes, including the length header. */
+#define TVM_CRT_MAX_PACKET_SIZE_BYTES 32
+
+/* Maximum supported string length in dltype, e.g. "int8", "int16", "float32" */
+#define TVM_CRT_MAX_STRLEN_DLTYPE 10
+
+/* Maximum supported string length in function names */
+#define TVM_CRT_MAX_STRLEN_FUNCTION_NAME 64
+
+/* Maximum supported string length in parameter names */
+#define TVM_CRT_MAX_STRLEN_PARAM_NAME 8
+
+#endif // MODULES_TVM_CRT_CONFIG_H_
diff --git a/zephyr/module.yml b/zephyr/module.yml
new file mode 100644
index 000000000..3fd601211
--- /dev/null
+++ b/zephyr/module.yml
@@ -0,0 +1,4 @@
+name: tvm
+build:
+  cmake-ext: True
+  kconfig-ext: True
