diff --git a/include/tvm/runtime/crt/packed_func.h b/include/tvm/runtime/crt/packed_func.h
index 83d961baf..c5cd25e00 100644
--- a/include/tvm/runtime/crt/packed_func.h
+++ b/include/tvm/runtime/crt/packed_func.h
@@ -44,7 +44,7 @@ typedef struct TVMArgs {
 TVMArgs TVMArgs_Create(TVMValue* values, uint32_t* tcodes, uint32_t values_count);
 
 typedef struct TVMPackedFunc {
-  char name[200];
+  char name[TVM_CRT_MAX_STRLEN_FUNCTION_NAME];
   TVMFunctionHandle fexec;
   TVMArgs args;
   TVMArgs ret_value;
