diff --git a/src/runtime/crt/graph_executor/graph_executor.c b/src/runtime/crt/graph_executor/graph_executor.c
index 395a343cc..74d9b28d6 100644
--- a/src/runtime/crt/graph_executor/graph_executor.c
+++ b/src/runtime/crt/graph_executor/graph_executor.c
@@ -1065,7 +1123,7 @@ int TVMGraphExecutor_SetupOpExecs(TVMGraphExecutor* executor) {
   }
   for (nid = 0; nid < executor->nodes_count; nid++) {
     const TVMGraphExecutorNode* inode = executor->nodes + nid;
-    if (strcmp(inode->op_type, "null")) {
+    if (strcmp(inode->op_type, "null") && strcmp(inode->param.func_name, "__nop")) {
       DLTensorPtr args[TVM_CRT_MAX_ARGS];
       uint32_t args_count = 0;
       for (idx = 0; idx < inode->inputs_count; idx++) {
@@ -1094,7 +1152,10 @@ int TVMGraphExecutor_SetupOpExecs(TVMGraphExecutor* executor) {
       printf("tvm_op: creating %s with node_id=%d\n", inode->param.func_name, nid);
 #endif  // TVM_CRT_DEBUG
       TVMPackedFunc pf;
-      TVMGraphExecutor_CreateTVMOp(executor, &(inode->param), args, args_count, &pf);
+      status = TVMGraphExecutor_CreateTVMOp(executor, &(inode->param), args, args_count, &pf);
+      if (status) {
+        break;
+      }
       executor->op_execs[nid] = pf;
     } else {
       memset(&executor->op_execs[nid], 0, sizeof(TVMPackedFunc));
