From b0cd29d066ed8469f03f6351424c205ba839e13d Mon Sep 17 00:00:00 2001
From: Maciej Sobkowski <msobkowski@antmicro.com>
Date: Tue, 20 May 2025 11:54:15 +0200
Subject: [PATCH 09/12] instumentation: enable instrumentation in C++

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>
---
 CMakeLists.txt | 1 +
 1 file changed, 1 insertion(+)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 104232cabe8..17a4c0db9af 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -488,6 +488,7 @@ zephyr_link_libraries_ifndef(CONFIG_NATIVE_LIBRARY
 
 # @Intent: Enable function instrumentation injection at compile time
 zephyr_compile_options($<$<COMPILE_LANGUAGE:C>:$<TARGET_PROPERTY:compiler,func_instrumentation>>)
+zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:$<TARGET_PROPERTY:compiler,func_instrumentation>>)
 
 # Allow the user to inject options when calling cmake, e.g.
 # 'cmake -DEXTRA_CFLAGS="-Werror -Wno-deprecated-declarations" ..'
-- 
2.49.0

