diff --git a/modules/dlpack/CMakeLists.txt b/modules/dlpack/CMakeLists.txt
new file mode 100644
index 00000000000..814a4ea97f1
--- /dev/null
+++ b/modules/dlpack/CMakeLists.txt
@@ -0,0 +1,13 @@
+# Copyright (c) 2023-2024 Antmicro <www.antmicro.com>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_ZEPHYR_DLPACK_MODULE)
+  set(DLPACK_DIR ${ZEPHYR_CURRENT_MODULE_DIR})
+
+  zephyr_library()
+
+  zephyr_include_directories(
+    ${DLPACK_DIR}/include/
+  )
+endif(CONFIG_ZEPHYR_DLPACK_MODULE)
diff --git a/modules/dlpack/Kconfig b/modules/dlpack/Kconfig
new file mode 100644
index 00000000000..a14c9831805
--- /dev/null
+++ b/modules/dlpack/Kconfig
@@ -0,0 +1,7 @@
+# Copyright (c) 2023-2024 Antmicro <www.antmicro.com>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config ZEPHYR_DLPACK_MODULE
+        bool "Support for dlpack module"
+        default 0
